imgpkg/002-recursive-bundles/README.md

```diff
@@ -0,0 +1,280 @@
+copy | imported 4 images
+Succeeded
+```
+
+### From Repository to Repository
+
+#### Required changes
+
+##### Retrieve all images for all bundles
+
+The most significant change in this operation is that when we copy the images `imgpkg` will have to traverse all the bundles to collect all the images to copy.
+
+##### Bundles/Images deduped
+
+When copying the images/bundles ensure that `imgpkg` does not try to copy the same image multiple times
+
+#### Example
+
+Given that we create a bundle using the command 
+
+`imgpkg push -b ghcr.io/k14s/design-docs/simple-app-install-package -f imgpkg/002-recursive-bundles/examples/recursive-bundle`
+
+```=
+imgpkg copy -b ghcr.io/k14s/design-docs/simple-app-install-package --to-repo some-other.registry.io/recursive-bundle
+copy | exporting 4 images...
+copy | will export ghcr.io/k14s/design-docs/simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
+copy | will export ghcr.io/k14s/design-docs/utilities@sha256:47ae428a887c41ba0aedf87d560eb305a8aa522ffb80ac1c96a37b16df038e0f
+copy | will export ghcr.io/k14s/design-docs/simple-app-bundle@sha256:d211dd700949154e429d28661d01c99d53a38af0d5275842ccbf0bf6dbef8ca4
+copy | will export ghcr.io/k14s/design-docs/simple-app-install-package@sha256:77c97e82306fb2a616da0a78796db039aa76e5bac1508954d40c0c10c073e035
+copy | exported 4 images
+copy | importing 4 images...
+copy | importing ghcr.io/k14s/design-docs/simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0 -> some-other.registry.io/recursive-bundle@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0...
+copy | importing ghcr.io/k14s/design-docs/utilities@sha256:ghcr.io/k14s/design-docs/utilities@sha256:47ae428a887c41ba0aedf87d560eb305a8aa522ffb80ac1c96a37b16df038e0f -> some-other.registry.io/recursive-bundle@sha256:ghcr.io/k14s/design-docs/utilities@sha256:47ae428a887c41ba0aedf87d560eb305a8aa522ffb80ac1c96a37b16df038e0f...
+copy | importing ghcr.io/k14s/design-docs/simple-app-bundle@sha256:d211dd700949154e429d28661d01c99d53a38af0d5275842ccbf0bf6dbef8ca4 -> some-other.registry.io/recursive-bundle@sha256:d211dd700949154e429d28661d01c99d53a38af0d5275842ccbf0bf6dbef8ca4...
+copy | importing ghcr.io/k14s/design-docs/simple-app-install-package@sha256:77c97e82306fb2a616da0a78796db039aa76e5bac1508954d40c0c10c073e035 -> some-other.registry.io/recursive-bundle@sha256:77c97e82306fb2a616da0a78796db039aa76e5bac1508954d40c0c10c073e035...
+copy | imported 4 images
+Succeeded
+```
+
+## Pull a bundle
+
+### Required changes
+
+#### Change folder structure
+
+When downloading a bundle that contains other bundles to disk using the `pull` command will work as currently, except for that it will download the nested bundles into a hidden folder called `.bundles`.
+
+##### Folder structure
+
+The structure of the output folder will be
+```
+$ tree -a recursive-bundle
+.
+├── .bundles
```

Comment:
i think it might be better to use `.imgpkg/bundles` instead of `.bundles` to:
- indicate that this is part of imgpkg functionality
- reduce out interface from fs perspective

Reply:
>indicate that this is part of imgpkg functionality

I agree that this is definitely internal functionality, nevertheless, this is something that the deployer will have to be aware of.

Let's imagine you are trying to deploy an application with a recursive bundle. You will have to deploy the configuration in the root of the bundle and also go into `.imgpkg/bundles/*` in order to deploy these other bundles configuration. In the end, we would be hiding something that the deployer would have to use.

Through these lenses, I would prefer to extract the contents of each bundle to the root folder, but unfortunately, that would create the possibility of collisions of file/folder names.
I feel like the compromise of pushing these folders into `.bundles` gives the deployer a clear signal that there is something else that needs to be installed.

Reply:
> I agree that this is definitely internal functionality

i did not mean that functionality is internal. i meant that it's very much user facing but belongs to imgpkg as opposed to user included files in the bundle.

> In the end, we would be hiding something that the deployer would have to use.

deployer needs to know about images.yml. i see it as opposite to hiding -- placing it in the expected location (location that is related to imgpkg)

> gives the deployer a clear signal that there is something else that needs to be installed.

the way im seeing it is that if user of bundle requested to download referenced bundles then they clearly know to go and use them. if they did not request is as part of pull command then there is nothing for them to do. now that i think about it i didnt see any new flag to indicate to pull referenced bundles...

Reply:
The proposal was updated to reflect this suggestion